File: problems/credit.py
import cs50

credit = cs50.get_string("Number: ")

# выстраивает цифры карты в обратном направлении
copy_credit = credit[::-1]
# умножает на два каждую вторую цифру, начиная с цифры с индексом 1 (т.е. со второй)
#every_other_sum = [int(x) * 2 for x in copy_credit[1::2]]
# умножает на два кажду вторую цифру и результат умножения складывает между собой и суммирует все получившееся
# например цифра в карте была 9, *2 = 18, 1+8
# int от деления 18 на 10 = 1, остаток от деления 18 на 10 = 8
every_other_sum = sum([(int(x) * 2 // 10 + int(x) * 2 % 10) for x in copy_credit[1::2]])
# сумма цифр карты, начиная с цфиры с индексом 0 (т.е. с первой)
first_sum = sum([int(x) for x in copy_credit[0::2]])
total_sum = first_sum + every_other_sum

if total_sum % 10 == 0:
    if len(credit) == 15 and credit[0:2] in ['34', '37']:
        print('AMEX')
    elif len(credit) == 16 and 51 <= int(credit[0:2]) <= 55:
        print('MASTERCARD')
    elif len(credit) in [13, 16] and credit[0] == '4':
        print('VISA')
    else:
        print('INVALID')
else:
    print('INVALID')

File: problems/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    long  card_number = get_long("Card number?\n");
    long cn = card_number;
    int v = 0;
    int i = 0;
    int rm;
    do
    {
        rm = cn % 10;
        i = i + 1;
        if (i % 2 == 0)
        {
            int m = 2 * rm;
            rm = 0;
            do
            {
                int r = m % 10;
                rm = rm + r;
                m = m / 10;
            }
            while (m > 0);

        }
        v = v + rm;
        cn = cn / 10;

    }
    while (cn > 0);

    int x = card_number / pow(10, (i - 1));
    int y = card_number / pow(10, (i - 2));


    if (v % 10 == 0 && i == 15 && (y ==  37 || y == 34))
    {
       printf("AMEX\n");
    }
    else if (v % 10 == 0 && i == 16 && (y == 51 || y == 52 || y == 53 || y == 54 || y == 55))
    {
        printf("MASTERCARD\n");
    }
    else if (v % 10 == 0 && x == 4 && (i == 16 || i == 13))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

}